### PR TITLE
[Kernel] Add moe_gating_top_k operator support for Ascend NPU

### PR DESCRIPTION
**Pull request by ZCG12345**
### What this PR does / why we need it?

1.replace moe_gating_top_k from torch_npu with custom op
2.enable the  renorm function of moe_gating_top_k in softmax scenerio

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
No need test

- vLLM version: v0.13.0
- vLLM main: https://github.com/vllm-project/vllm/commit/7157596103666ee7ccb7008acee8bff8a8ff1731

### CODE REVIEW

File: tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+        for (group_select_mode, renorm, norm_type, group_count, k, dim0,
+             dim1) in param_product:
+
+            group_count = 8
+            k = 5
+            if k > dim1 // group_count:
+                continue
+
+            # ---- 构造输入 ----
```
Comment (zzzzwwjj) [inline tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py]: Please change to English.

File: tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+# random.seed(450)
+seed = 45
+random.seed(seed)
+numpy.random.seed(seed)
+torch.manual_seed(seed)
+
+
+def softmax_func(x, axis=None):
+    # print("before softmax:",x)
```
Comment (zzzzwwjj) [inline tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py]: Please remove useless code instead of comment out it.

**Comment by realliujiaxu** [comment]
please update PR title to include these details:
1. replace `moe_gating_top_k` from torch_npu with custom op
2. add renorm to `moe_gating_top_k`

File: tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                if bias_tensor is not None else None,
+            )
+
+            # ---- Output current case information ----
+            print(
+                f"[Case] x=({dim0},{dim1}), k={k}, "
+                f"group_count={group_count}, select_mode={group_select_mode}, "
+                f"norm_type={norm_type}, renorm={renorm},"
+                f"k_group={k_group}")
+            print(y_npu.shape, expert_idx_npu.shape, out_npu.shape)
```
Comment (realliujiaxu) [inline tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py]: please check if it is necessary to print these tensors and metadata.

File: vllm_ascend/ops/fused_moe/experts_selector.py [hunk 4]
```diff
@@ -209,26 +208,33 @@ def _select_experts_with_fusion_ops(
...
+        renorm=renorm,  # 0: softmax->topk(fix); 1: topk->softmax
         norm_type=norm_type,  # 0: softmax; 1: sigmoid
-        # out_flag=False, # todo new api; should the third output be output
-        # y2_flag=False, # old api; should the third output be output
-        routed_scaling_factor=routed_scaling_factor,
-        eps=float(1e-20))
-    if scoring_func == "softmax":
-        topk_weights = _renormalize_topk_weights(topk_weights, renormalize)
+        out_flag=False,  # todo new api; should the third output be output
+        routed_scaling_factor=1,
```
Comment (realliujiaxu) [inline vllm_ascend/ops/fused_moe/experts_selector.py]: 1. why the `routed_scaling_factor` is hardcoded to `1`?
2. since default value of `routed_scaling_factor` is `1.0`, this line should be deleted.

File: vllm_ascend/ops/fused_moe/experts_selector.py [hunk 5]
```diff
@@ -209,26 +208,33 @@ def _select_experts_with_fusion_ops(
 
     topk_group = topk_group if topk_group is not None else 1
     num_expert_group = num_expert_group if num_expert_group is not None else 1
+    renorm = int(renormalize)
+    if scoring_func == "sigmoid" and renorm == 0:
+
+        raise ValueError(
```
Comment (realliujiaxu) [inline vllm_ascend/ops/fused_moe/experts_selector.py]: `NotImplementedError` maybe better

File: csrc/torch_binding.cpp [hunk 6]
```diff
@@ -1218,10 +1218,65 @@ std::tuple<at::Tensor, at::Tensor, at::Tensor, at::Tensor> npu_moe_init_routing_
...
+                    k,                   // k
+                    k_group,              // k_group
+                    group_count,          // group_count
+                    group_select_mode,   // group_select_mode
+                    renorm,              // renorm
+                    norm_type,            // norm_type
+                    out_flag,             // out_flag
+                    routed_scaling_factor, // routed_scaling_factor
+                    eps,                 // eps
+                    y,                // input_y (注意：这里应该是 yOut)
```
Comment (realliujiaxu) [inline csrc/torch_binding.cpp]: change this to English @@ZCG12345

File: tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py (hunk 3 above)
Comment (wangxiyuan) [reply tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py]: no `print` in code. Please

File: vllm_ascend/ops/fused_moe/experts_selector.py [hunk 7]
```diff
@@ -209,26 +208,33 @@ def _select_experts_with_fusion_ops(
 
     topk_group = topk_group if topk_group is not None else 1
     num_expert_group = num_expert_group if num_expert_group is not None else 1
+    renorm = int(renormalize)
+    if scoring_func == "sigmoid" and renorm == 0:
```
Comment (whx-sjtu) [inline vllm_ascend/ops/fused_moe/experts_selector.py]: For this situation, we need to go into small ops instead of fused op if it is currently not supported. So please move this checking to https://github.com/vllm-project/vllm-ascend/blob/8ffe3f5d7872b58a6a69825673cb0cb1d37e9193/vllm_ascend/ops/fused_moe/experts_selector.py#L102 and make this fucntion return False in this scenario.

**Review by whx-sjtu — CHANGES_REQUESTED** [review]

File: tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+            y /= (numpy.sum(y, axis=-1, keepdims=True) + eps)
+        y *= routed_scaling_factor
+        if y2_flag:
+            y2 = original_x
+        else:
+            y2 = None
+        y = torch.tensor(y, dtype=dtype)
+        return y, indices.astype(numpy.int32), y2
+
+    def test_npu_moe_gating_topk_multi(self, device="npu"):
```
Comment (realliujiaxu) [inline tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py]: Please modify this test case using Python's parameterized testing：
```
@pytest.mark.parametrize("group_select_modes",  [0, 1]))
@pytest.mark.parametrize("group_counts", [1, 8])
@pytest.mark.parametrize("xxx", [xxx])
def test_npu_moe_gating_topk_multi(self, group_select_mode, group_count,...):
```

**Review by realliujiaxu — CHANGES_REQUESTED** [review]
please update test case `test_npu_moe_gating_topk_multi` and check `routed_scaling_factor`

File: tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+            bias_tensor = torch.tensor(bias, dtype=torch.float32)
+            # bias_tensor = None
+            k_group = random.randint(1, group_count)
+            out_flag = False
+            routed_scaling_factor = 1.0
+            eps = 1e-20
+
+            # ---- error debug ----
+
+            # k = 4
```
Comment (zzzzwwjj) [inline tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py]: Please remove useless comment in code.

File: tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        k_group=k_group,
+        group_count=group_count,
+        group_select_mode=group_select_mode,
+        renorm=renorm,
+        norm_type=norm_type,
+        out_flag=out_flag,
+        routed_scaling_factor=routed_scaling_factor,
+        eps=eps,
+        bias_opt=bias_tensor.npu() if bias_tensor is not None else None,
+    )
```
Comment (realliujiaxu) [inline tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py:195]: please add `assert` to make sure output is the same

**Review by whx-sjtu — APPROVED** [review]

File: tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py (hunk 10 above)
Comment (ZCG12345) [reply tests/e2e/nightly/single_node/ops/singlecard_ops/test_npu_moe_gating_top_k.py:195]: OK Thanks jiaxu my bro ,If there are anything else need to be corrected you can tell me . 
Best wishes

**Comment by realliujiaxu** [comment]
please change all comment to English. Apart from that, this PR is LGTM.

**Comment by ZCG12345** [comment]
> please change all comment to English. Apart from that, this PR is LGTM.

yes jiaxu bro ,already change

File: vllm_ascend/ops/fused_moe/experts_selector.py (hunk 7 above)
Comment (ZCG12345) [reply vllm_ascend/ops/fused_moe/experts_selector.py]: OK ,Thankyou advise my bro

**Review by zzzzwwjj — APPROVED** [review]

**Review by realliujiaxu — APPROVED** [review]
